lib/cloud-operation/cloud-operation-v2.ts

```diff
@@ -0,0 +1,94 @@
+import { CloudOperationBase } from "./cloud-operation-base";
+import { CloudOperationMessageTypes, CloudCommunicationEvents } from "../constants";
+
+class CloudOperationV2 extends CloudOperationBase implements ICloudOperation {
+	private static readonly WAIT_TO_START_TIMEOUT: number = 10 * 60 * 1000;
+
+	private communicationChannel: ICloudCommunicationChannel;
+	private waitToStartTimeout: NodeJS.Timer;
+	private waitResultPromise: Promise<ICloudOperationResult>;
+
+	constructor(protected id: string,
+		protected serverResponse: IServerResponse,
+		private $nsCloudCommunicationChannelFactory: ICloudCommunicationChannelFactory) {
+		super(id, serverResponse);
+
+		this.communicationChannel = this.$nsCloudCommunicationChannelFactory.create(serverResponse.communicationChannel, this.id);
+	}
+
+	public async sendMessage<T>(message: ICloudOperationMessage<T>): Promise<void> {
+		this.isInitialized();
+		await this.communicationChannel.sendMessage(message);
+	}
+
+	public async cleanup(exitCode?: number): Promise<void> {
+		if (this.waitToStartTimeout) {
+			clearTimeout(this.waitToStartTimeout);
+			this.waitToStartTimeout = null;
+		}
+
+		await this.communicationChannel.close(exitCode);
+	}
+
+	protected initCore(): Promise<void> {
+		return new Promise<void>(async (resolve, reject) => {
+			this.waitToStartTimeout = setTimeout(() => {
+				clearTimeout(this.waitToStartTimeout);
+				this.waitToStartTimeout = null;
+				if (!this.status) {
+					return reject(new Error("Communication channel init timed out."));
+				}
+			}, CloudOperationV2.WAIT_TO_START_TIMEOUT);
+
+			const closeHandler = (code: number) => {
+				this.cleanup(code);
+				reject(new Error(`Communication channel closed with code ${code}`));
+			};
+			try {
+				this.communicationChannel.once(CloudCommunicationEvents.CLOSE, closeHandler);
+				await this.communicationChannel.connect();
+				this.status = CloudOperationV2.OPERATION_IN_PROGRESS_STATUS;
```

Comment:
[Could] Not a big deal, but can we have this.status and this.waitResultPromise setters on subsequent lines (:

Reply:
![especially for you](https://media.giphy.com/media/l0ErOholJjSmFlMFG/giphy.gif)